### ISSUE TITLE
Support NUL terminated input/output for selecting files with arbitrary characters (eg newlines)

### ISSUE THREAD

**Issue by HaleTom** [peco/peco#419]
# Feature request
Would you consider supporting NUL (0x00) terminated input/output for selecting files with arbitrary characters (eg newlines)?

# Discussion

`man xargs` says:

>       -0, --null
>            Input items are terminated by a null character instead of by whitespace, and the quotes and backslash are not special (every character is taken literally).   Disables  the
>            end of file string, which is treated like any other argument.  Useful when input items might contain white space, quote marks, or backslashes.  The GNU find -print0 option
>            produces input suitable for this mode.

Would you consider supporting adding a `-0` option which expects NUL-terminated input and writes NUL-terminated output?  

**Example**
```
find . -type f -print0 | peco -0 | xargs -0 myprog
```

**Comment by lestrrat** [comment]
`-0` seems vague, maybe `--print-null` or something?

I confess I've never used `xargs -0`, so would love to have a test (possibly pseudocode) that demonstrates exactly what the behavior should be.

Also, peco has a `--null` mode. I think we'd have to define what the expected behavior is when you do something like

```
find . -type f -print0 | peco --null --print-null | xargs -0 myprog
```

Would be nice if that semantics is incorporated in the test/pseudocode above.

**Comment by lestrrat** [comment]
I'm pretty sure I understand, but again as my original reply says, I really would like a test case so I can code against it. For example, said test only needs to provide me with a few given input "X", we expect "Y", like:

```go
func TestThisAwesomeFeature(t *testing.T) {
   tests := []struct {
       src string
       expected
   } {
       { src: "...", expected: "..." },
        ....
   }
}
```

This is crucial to make align everybody's expectations

**Comment by bastienbc** [comment]
Personally, I would expect `peco` to act the same as `grep`, but I can understand that someone may argue something else.

Also I noticed when I was building a test case that `grep` doesn't change back `\0` to a newline, unlike `fzf`. I like that because this tool shouldn't expect that it should change back (back from what?).

So here is a test case for you:
```
$ echo -en 'AAB\0AOB\0AAP\n\tOPT\0PCO\nAAD\0AGO\0BCA\0BAA' | sort -z | grep -z 'AA' | tr '\0' '\n'
AAB
AAP
        OPT
BAA
PCO
AAD
```

And I expect `peco` to give the same output if I select every line from:
```
$ echo -en 'AAB\0AOB\0AAP\n\tOPT\0PCO\nAAD\0AGO\0BCA\0BAA' | sort -z |  peco --null-input --initial-filter Regexp --query 'AA' | tr '\0' '\n'
```

Since `peco` is interactive I don't know how to write an automatic test script for that, and I don't know where to start with your Go code.

**Comment by lestrrat** [comment]
@bastienbc I don't need a real working test: I need a few variations on input text and output text, exactly as written in https://github.com/peco/peco/issues/419#issuecomment-382892337

**Comment by lestrrat** [comment]
> For the test case, how would you say that with an input of foo\nbar\0 should be one item of foo\nbar > rather than two items of foo and bar?

I personally only use peco for very limited stuff, and don't use any of this (and will in the future try to avoid passing this sort of stuff through peco with all of my might), so really, the exact semantics are up to you guys here.

My stance is: give me concrete test cases to work against and a good enough rationale (and while I might tweak a few bits here and there) I'm just going to go with the option that has most tests and that sounds most reasonable :-)

**Comment by bastienbc** [comment]
@HaleTom You are right about grep, i just glanced at the options without really reading.

But the point of that option is to keep `foo\nbar` as a single object in my opinion.
Maybe having `-z` as a shortcut for `-d \0` ? That would not change the default behaviour, but would be more flexible. But that may wait for later :-)

@lestrrat I don't know if that is exactly what you are waiting but here are some tests cases:
```golang
func TestNullStringSeparator(t *testing.T) {
   tests := []struct {
       src string
       expected []string
   } {
{ src: "", expected: []string{""} },
{ src: "foo", expected: []string{"foo"} },
{ src: "foo\0bar", expected: []string{"foo","bar"} },
{ src: "foo\nbar", expected: []string{"foo\nbar"} },
{ src: "foo\nbar\0third", expected: []string{"foo\nbar","third"} },
{ src: "foo\0bar\0third", expected: []string{"foo","bar","third"} },
{ src: "foo\0bar\nthird\0average\nexample\ncase\0nothing\0foo", expected: []string{
"foo",
"bar\nthird",
"average\nexample\ncase",
"nothing",
"foo" },
   }
}
```

I could try to do a PR maybe.

**Comment by lestrrat** [comment]
@bastienbc works for me. I can't promise if I can start working on it immediately, but it's now on me, so I only have myself to blame... :)

**Comment by HaleTom** [comment]
I couldn't see -d in the Readme.md... did I miss it?

Is the usage of --null in the readme supposed to point to a directory?

**Comment by bastienbc** [comment]
@HaleTom It's just an idea on how to implement `-z`. `-d` doesn't exist. Yet.

**Comment by lestrrat** [comment]
hopefully fixed by #577?
